plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -385,14 +385,23 @@ public synchronized void setTableStatistics(HiveIdentity identity, Table table,
     // TODO: Allow updating statistics for 2 tables in the same transaction
     public synchronized void setPartitionStatistics(HiveIdentity identity, Table table, Map<List<String>, PartitionStatistics> partitionStatisticsMap)
     {
-        setExclusive((delegate, hdfsEnvironment) ->
-                partitionStatisticsMap.forEach((partitionValues, newPartitionStats) ->
-                        delegate.updatePartitionStatistics(
-                                identity,
-                                table.getDatabaseName(),
-                                table.getTableName(),
-                                getPartitionName(table, partitionValues),
-                                oldPartitionStats -> updatePartitionStatistics(oldPartitionStats, newPartitionStats))));
+        setExclusive((delegate, hdfsEnvironment) -> {
+            ImmutableList.Builder<CompletableFuture<?>> executeUpdateFutures = ImmutableList.builder();
+            for (Map.Entry<List<String>, PartitionStatistics> partitionEntry : partitionStatisticsMap.entrySet()) {
+                executeUpdateFutures.add(CompletableFuture.runAsync(() -> {
+                    delegate.updatePartitionStatistics(
+                            identity,
+                            table.getDatabaseName(),
+                            table.getTableName(),
+                            getPartitionName(table, partitionEntry.getKey()),
+                            oldPartitionStats -> updatePartitionStatistics(oldPartitionStats, partitionEntry.getValue()));
+                }, updateExecutor));
```

Comment:
This looks fine. I was a bit concerned at first if the method is not called from context which already uses `updateExecutor`. The we could have a deadlock here if thread pool in executor got depleted. But it looks it is not the case.